feat: Add Created date to Auto-Suggest menu

ldanet opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

This adds the created date ➕ suggestion, which inserts the ➕ symbol with the current date when selected.
I'm not sure what the exact implementation should look like but I'm happy to take feedback and update the PR accordingly!

## Motivation and Context

Fixes #1743 .

## How has this been tested?

I built `main.js`, replaced the existing `main.js` in the plugin folder in my test vault.
I then manually tried to create tasks and edit existing ones to view the suggestions and select them in various orders.
I also did the smoke test in the sample vault.

## Screenshots (if appropriate)
<img width="313" alt="Screenshot 2023-04-15 at 12 58 33" src="https://user-images.githubusercontent.com/6631995/232239577-eef5a165-cbe8-43e5-a9c7-74ad1d843eed.png">
After selecting "➕ created today":
<img width="419" alt="Screenshot 2023-04-15 at 12 58 44" src="https://user-images.githubusercontent.com/6631995/232239613-c480e853-8e89-42d2-972a-c6711b944279.png">

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [ ] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [x] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [x] **Documentation** (prefix: `docs` - improvements to any documentation content **for users**)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))
- [ ] **Contributing Guidelines** (prefix: `contrib` - any improvements to documentation content **for contributors** - see [Contributing to Tasks](https://publish.obsidian.md/tasks-contributing/))

Internal changes:

- [ ] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [ ] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [x] My change requires a change to the documentation.
- [x] I have [updated the documentation](https://publish.obsidian.md/tasks-contributing/Documentation/About+Documentation) accordingly.
- [x] My change has adequate [Unit Test coverage](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://publish.obsidian.md/tasks-contributing) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://publish.obsidian.md/tasks-contributing)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

ldanet commented on docs/Getting Started/Auto-Suggest.md:
```diff
@@ -59,7 +59,7 @@ Here is a more detailed walk through of the creation of a new task, which can be
 7. Now our task is complete:
 
-    ![auto-suggest-menu-after-due-emoji](../images/auto-suggest-menu-after-due-emoji.png)
+    ![auto-suggest-menu-after-due-emoji](../images/auto-suggest-menu-after-due-date.png)
```
I noticed the image was repeated in the documentation, so I took the opportunity to change it to what seems like the correct one instead.

claremacrae replied:
OK, thanks. Oh, that's much better.

The alt-text for screen-readers was rubbish before, but please could you update it as well, to at least make It consistently rubbish? 🤣

ldanet replied:
I did not notice the alt text! 😂 Fixing it!

ldanet commented on tests/Suggestor/Suggestor.test.ts:
```diff
@@ -2,16 +2,31 @@
+const mockDate = new Date(moment('2022-07-11 15:00').valueOf());
+
+jest.spyOn(chrono, 'parseDate').mockImplementation(
+    (text, _, options) => chrono.en.casual.parseDate(text, mockDate, options)!,
```
`chrono.parseDate` is a shortcut to `chrono.en.casual.parseDate` so the result is the same.

ldanet commented on src/Suggestor/Suggestor.ts:
```diff
@@ -98,6 +98,14 @@ function getPossibleComponentSuggestions(
             displayText: `${symbols.scheduledDateSymbol} scheduled date`,
             appendText: `${symbols.scheduledDateSymbol} `,
         });
+    if (!line.includes(symbols.createdDateSymbol)) {
+        const parsedDate = DateParser.parseDate('today', true);
+        const formattedDate = parsedDate.format(TaskRegularExpressions.dateFormat);
+        suggestions.push({
+            displayText: `${symbols.createdDateSymbol} created today (${formattedDate})`,
+            appendText: `${symbols.createdDateSymbol} ${formattedDate} `,
+        });
+    }
```
I tentatively put it after date-like suggestions and before priority, but I'm happy to move it!
Also, all the date suggestions include the computed date in parentheses so I replicated the pattern here too, but I'm not actually sure it's very useful.

ldanet commented on src/Suggestor/Suggestor.ts:
```diff
@@ -183,7 +191,7 @@ function addDatesSuggestions(
         // a max number. We want the max number to be around half the total allowed matches, to also allow
         // some global generic matches (e.g. task components) to find their way to the menu
         const minMatch = 1;
-        const maxGenericSuggestions = 5;
+        const maxGenericSuggestions = settings.autoSuggestMaxItems - 1;
```
I wanted this variable to be derived from the settings max suggestions, so the tests could run without being limited to an arbitrary 5. I changed this to max items - 1 because the default max suggestions is 6 and it was set to 5, but now that I'm seeing it in the context of the PR, I notice the comment above says it should be half of the matches! Let me update this 😅

ldanet replied:
I updated it to be half of the maximum suggestions, but then I thought a bit more about it... 

Once you've added a priority, a created date and a due date, there are only 3 general suggestions left (start, scheduled, recurrence). I you then want to add, say, a recurrence, first you select it (only 2 general suggestions left), and then it only suggests max / 2 recurrence suggestions, which is 3 with the default setting of 6 suggestions. That means that we're only showing 3 + 2 = 5 items total, when the user wanted 6 suggestions. That means we're not being as helpful as we could be! It gets more obvious if the user has selected to see more suggestions than the default.

With a little refactor, that is, moving the general suggestions generation first so we know how many there are (and adding them later), we can determine how much space we actually have for the date/recurrence suggestions. I have it working, but it's a bit outside the scope of the original issue though. Would you like me to add this to the PR? This is what it looks like: https://github.com/ldanet/obsidian-tasks/commit/e45603ec4cbf2136f01651e8a0c1777204a971f0

claremacrae replied:
Thank you for this suggestion. Sorry but I'm not going to have any cycles to look at this until Saturday at the earliest...

ldanet replied:
No rush at all!

claremacrae commented:
Wow. Thank you very much @ldanet. I'm reviewing several PRs at the moment, and will test this out as soon as I can.

I really appreciate the contribution! 🎉

ldanet commented on tests/Suggestor/Suggestor.test.ts:
```diff
@@ -22,8 +37,11 @@ describe.each([{ name: 'emoji', symbols: DEFAULT_SYMBOLS }])("auto-complete with
             { displayText: `${dueDateSymbol} due date`, appendText: `${dueDateSymbol} ` },
             { displayText: `${startDateSymbol} start date`, appendText: `${startDateSymbol} ` },
             { displayText: `${scheduledDateSymbol} scheduled date`, appendText: `${scheduledDateSymbol} ` },
+            {
+                displayText: `${createdDateSymbol} created today (2022-07-11)`,
+                appendText: `${createdDateSymbol} 2022-07-11 `,
+            },
```
I had to figure out a way to test dates in this file for this!

ldanet commented on tests/Suggestor/Suggestor.test.ts:
```diff
@@ -2,16 +2,31 @@
 window.moment = moment;
 
+// Set predictable date for all tests in this file
+const mockDate = new Date(moment('2022-07-11 15:00').valueOf());
```
I arbitrarily put `15:00` so that the date wasn't midnight. At midnight, `chrono.parseDate` is happy with "Monday" the same day being in the future, so it was failing Mondays in the big test below.

claremacrae commented:
Whilst you were working on this, I edited a line in Auto-Suggest.md that you correctly deleted.

How comfortable are you with Git and merging in the latest from `main`?

If you are 100% comfortable, please could you merge in main, so that I can tell GitHub to go ahead and run the tests here?

ldanet commented:
> Whilst you were working on this, I edited a line in Auto-Suggest.md that you correctly deleted.
> 
> How comfortable are you with Git and merging in the latest from `main`?
> 
> If you are 100% comfortable, please could you merge in main, so that I can tell GitHub to go ahead and run the tests here?

I'm very comfortable with git! I'll update and fix any conflicts!

claremacrae commented:
Any chance you could also edit this bit of the docs, please, to add to the ways that Create date can be - erm - created:

https://publish.obsidian.md/tasks/Getting+Started/Dates#Created+date

(You've probably seen already but editing the `docs/` folder as a vault in Obsidian is a great way of editing the docs, for ease of linking - and see how the docs will look when published)

claremacrae commented:
Apologies if you've done this already - I need to get some sleep ....

But I wonder, should it only offer to add Created date via auto-suggest if the created field is turned on in settings?

ldanet commented:
> Apologies if you've done this already - I need to get some sleep ....
> 
> But I wonder, should it only offer to add Created date via auto-suggest if the created field is turned on in settings?

That's definitely technically possible, but I'm not sure if it would be more helpful or more annoying from a user standpoint 🤔 

Would there be a case where users don't want the plugin to include the creation date automatically but would still like to add it manually if it's easy to do? Or can we safely assume that if they don't have enabled "Set created date on every task", they don't want to record creation dates at all? I guess the completed date is the similar in the sense that it's not easy to record the completion date if you do not enable the setting to have it automatically done either!

I personally don't mind either way, I have the setting enabled in my own vault 😄
I can add the check if you'd like.

claremacrae commented:
FYI When I smoke-test this, I will update any screenshots that have changed with this PR, and push to your branch.

Hope you don't mind. It saves me doing more work creating a PR or manual commit separately...

claremacrae commented on docs/images/auto-suggest-menu-after-due-emoji.png:
I feel that the reduction from 5 suggested dates to 3 is a problem, since once there is a date emoji, the guaranteed next thing a user will do is add a date.

ldanet replied:
I agree and using it, I wondered why property suggestions would show up at all at this point... I feel like we should do away with the limited number of date suggestions altogether. But as it currently stands, if we do that, the general task property suggestions wouldn't show up at all, even if they match text the user enters better than date suggestions. I'm going to put the maximum back the way it was in order to proceed with this PR, but I'm very very tempted to then create a new one to improve the suggestion algorithm in order to:
- show as many date/recurrence suggestions as we can without stopping the other suggestions from showing up if needed
- When a date is what is being added, create date suggestions on the fly if the user types a valid date shortcut that isn't in the set of default date suggestions?
- Refactor the task property suggestions into their own function that works the same as the others so the code is easier to navigate.

claremacrae replied:
>  and using it, I wondered why property suggestions would show up at all at this point...

(meaning, when the user has just selected a date emoji for example, there is no point offering them other emojis, until they have chosen a date...)

Yes, I do agree with that.

I'm very happy with what you suggest above... Thank you.

claremacrae reviewed (commented):
As promised I recreated the images.

First, though, I ran through the walk-through with Tasks 3.1.0 - and was able to reproduce all images.

And then tried again with the code in this PR and saved new images.

Trivia: I don't like the new border under the menu - it's uglier in screenshots - but will do for now.

claremacrae commented on docs/images/auto-suggest-menu-after-due-date.png:
For the record, the screenshots were created with this setting:

![image](https://user-images.githubusercontent.com/4840096/231577943-bf9978bd-79c3-453a-a7eb-10aeafe22357.png)

ldanet commented on docs/images/auto-suggest-menu-initial-menu.png:
Moving the created date to the bottom of the general component suggestions 👍

claremacrae replied:
> Moving the created date to the bottom of the general component suggestions 👍

@ldanet Please be more explicit. I do not understand your point.

ldanet replied:
I meant the created date now comes after priority and recurrence! I've started to name those types and suggestions "task property" suggestions and I might update the variable name to that in a future PR. Words are hard! 😄

ldanet commented:
It's updated and ready to review again. I moved the new created date suggestion to the bottom of the task property suggestions and reverted the change to the number of date suggestions. I also updated the screenshots in the documentation. ✨

claremacrae reviewed (approved):
Brilliant. Thank you very much.

I'm going to merge, and then experiment with making the new mocking code clear its value at the end of the test.... This should be written up in the Contributing guide, but is not yet.

claremacrae commented:
@ldanet FYI I am doing a little refactoring to enable the skipped test in `Suggestor.test.ts` having discovered that it has been failing for a little while, and I never knew because it was skipped....

I'm making it so that maxGenericSuggestions can be passed in to `makeDefaultSuggestionBuilder()`.

claremacrae commented:
> @ldanet FYI I am doing a little refactoring to enable the skipped test in `Suggestor.test.ts` having discovered that it has been failing for a little while, and I never knew because it was skipped....
> 
> I'm making it so that maxGenericSuggestions can be passed in to `makeDefaultSuggestionBuilder()`.

Cancel that. It would increase the conflicts with #1824.

claremacrae commented:
I think we should probably move this to a new issue, so comments don't get lost in this one... @esm7 would you be able to do that please?

claremacrae commented:
> I was originally about to do just that, but I thought that if what I see as inconsistent and unexpected behavior is actually intentional, there's no point to open a new issue for it. Is this the case or can it be considered an issue?

That's a fair point.

### Earlier discussion (from memory)

We debated above whether to make the `Created` field optional in Auto Suggest, depending on the user's setting.

In the end, I opted for sticking with the original (current) proposed implementation, to then see what feedback there is from users.

Also, there is also discussion above about refining the parameters used to control how many items - and what sort - are displayed in the auto suggest menu.

I don't remember any discussion about what happens when you specifically type `today`.

### A possible workaround

Also, @esm7 - what happens when you increase the number of suggestions in settings? Do you get the behaviour you want?

### Created and today

I do think that it is logical to make the Created suggestion add today's date. It will be rare for someone to want a different date.

### What next

If it's annoying, it doesn't matter if it was intentional or not... It's still valid feedback - and worth logging to track it, I feel.

esm7 commented:
> > I was originally about to do just that, but I thought that if what I see as inconsistent and unexpected behavior is actually intentional, there's no point to open a new issue for it. Is this the case or can it be considered an issue?
> 
> That's a fair point.
> ### Earlier discussion (from memory)
> 
> We debated above whether to make the `Created` field optional in Auto Suggest, depending on the user's setting.
> 
> In the end, I opted for sticking with the original (current) proposed implementation, to then see what feedback there is from users.
> 
> Also, there is also discussion above about refining the parameters used to control how many items - and what sort - are displayed in the auto suggest menu.
> 
> I don't remember any discussion about what happens when you specifically type `today`.
> ### A possible workaround
> 
> Also, @esm7 - what happens when you increase the number of suggestions in settings? Do you get the behaviour you want?
> ### Created and today
> 
> I do think that it is logical to make the Created suggestion add today's date. It will be rare for someone to want a different date.
> ### What next
> 
> If it's annoying, it doesn't matter if it was intentional or not... It's still valid feedback - and worth logging to track it, I feel.

My feedback is not related to the number of suggestions, on the contrary. As you can see in the video I posted above, "created today" can get triggered more than any other suggestion, and in contexts where no other suggestion is triggered.
I agree that the most sensible suggestion would be "created today", but in this case to be consistent with how the auto-suggest logic currently works, it needs to be triggered when the user starts typing "created". "today" on its own should not trigger a suggestion (as it never did), and definitely not after an emoticon of a different field.
If that's agreed upon I guess I can go ahead and open an issue.

claremacrae commented:
Please do report the bug, and if you have any suggestions for improvements then include them too. Thank you.
